### PR TITLE
feat: freelancer sign up with test relates #127

### PR DESCRIPTION
**Pull request by naghamabuwarda**

### CODE REVIEW

File: server/src/validation/index.ts [hunk 1]
```diff
@@ -76,6 +76,28 @@ const updateFreelancerValidation = yup.object().shape({
     .min(5, 'portfolio must be more than 5 characters long')
     .url('portfolio should be a valid URL'),
 });
+const freelancerValidate = yup.object().shape({
+  Title: yup
```
Comment (LinaYahya) [inline server/src/validation/index.ts:80]:
```suggestion
  title: yup
```

File: server/src/validation/index.ts [hunk 2]
```diff
@@ -76,6 +76,28 @@ const updateFreelancerValidation = yup.object().shape({
...
+    .string()
+    .required('Title is required'),
+  Major: yup
```
Comment (LinaYahya) [inline server/src/validation/index.ts:83]:
```suggestion
  major: yup
```

File: server/src/validation/index.ts [hunk 3]
```diff
@@ -76,6 +76,28 @@ const updateFreelancerValidation = yup.object().shape({
...
+    .string()
+    .required('Major is required'),
+  portfolio: yup
+    .string()
+    .optional()
+    .nullable()
+    .url(),
+  description: yup
```
Comment (LinaYahya) [inline server/src/validation/index.ts:91]:
```suggestion
  brief: yup
```

File: server/src/controllers/authentication.ts [hunk 4]
```diff
@@ -19,4 +20,26 @@ const login = async (req: Request, res: Response) => {
   res.cookie('token', token);
   return { status: 200, msg: 'logged in successfully' };
 };
-export default login;
+const freelancerSignUp = async (req: Request, res: Response) => {
+  const {
+    Title, Major, portfolio, description, image, id,
+  } = req.body;
+  const user: UserInstance | null = await User.findOne({ where: { id } });
+  if (!user) throw serverErrs.BAD_REQUEST('user id doen\'t exist');
```
Comment (LinaYahya) [inline server/src/controllers/authentication.ts:28]:
```suggestion
  if (!user) throw serverErrs.BAD_REQUEST('Something went wrong');
```

File: server/src/controllers/authentication.ts [hunk 5]
```diff
@@ -19,4 +20,26 @@ const login = async (req: Request, res: Response) => {
   res.cookie('token', token);
   return { status: 200, msg: 'logged in successfully' };
 };
-export default login;
+const freelancerSignUp = async (req: Request, res: Response) => {
+  const {
+    Title, Major, portfolio, description, image, id,
```
Comment (LinaYahya) [inline server/src/controllers/authentication.ts:25]:
```suggestion
   title, major, portfolio, brief, image, userId,
```

File: server/src/controllers/authentication.ts [hunk 6]
```diff
@@ -19,4 +20,26 @@ const login = async (req: Request, res: Response) => {
...
+  } = req.body;
+  const user: UserInstance | null = await User.findOne({ where: { id } });
+  if (!user) throw serverErrs.BAD_REQUEST('user id doen\'t exist');
+  const { name, role } = user;
+  const token = await generateToken({ userID: id, role, name });
```
Comment (LinaYahya) [inline server/src/controllers/authentication.ts:30]: First we have to `create` into freelancer model,
then we can generate token, specially that token `id` is freelancer id not `userId`

File: server/src/controllers/authentication.ts [hunk 7]
```diff
@@ -19,4 +20,26 @@ const login = async (req: Request, res: Response) => {
...
+
+  await freelancerValidate.validate(req.body);
```
Comment (LinaYahya) [inline server/src/controllers/authentication.ts:32]: validation goes first

File: server/src/controllers/authentication.ts [hunk 8]
```diff
@@ -19,4 +20,26 @@ const login = async (req: Request, res: Response) => {
...
+  const freelancer = await Freelancer.create({
+    image,
+    title: Title,
+    major: Major,
+    brief: description,
+    userId: id,
+    portfolio,
+  });
+  res.cookie('token', token);
+  return { status: 200, data: freelancer, msg: 'successful login' };
```
Comment (LinaYahya) [inline server/src/controllers/authentication.ts:43]:
```suggestion
  return { status: 201, data: freelancer, msg: 'successful login' };
```
And I think we don't need the whole data `name`, `role`, and freelancer `id` would be enough

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by mohammedagha27 — APPROVED** [review]

**Review by alishanti98 — APPROVED** [review]

**Review by shathakh — APPROVED** [review]

**Pull request merged**